ADAL/src/request/ADAuthenticationRequest+Broker.m

```diff
@@ -186,74 +185,82 @@ + (ADAuthenticationResult *)processBrokerResponse:(NSURL *)response
     }
     
     // NSURLComponents resolves some URLs which can't get resolved by NSURL
-    NSURLComponents* components = [NSURLComponents componentsWithURL:response resolvingAgainstBaseURL:NO];
+    NSURLComponents *components = [NSURLComponents componentsWithURL:response resolvingAgainstBaseURL:NO];
     NSString *qp = [components percentEncodedQuery];
     //expect to either response or error and description, AND correlation_id AND hash.
-    NSDictionary* queryParamsMap = [NSDictionary adURLFormDecode:qp];
-    
+    NSDictionary *queryParamsMap = [NSDictionary adURLFormDecode:qp];
+
+    NSString *keychainGroup = resumeDictionary[@"keychain_group"];
+
     if([queryParamsMap valueForKey:OAUTH2_ERROR_DESCRIPTION])
     {
-        return [ADAuthenticationResult resultFromBrokerResponse:queryParamsMap];
+        // In the case where Intune App Protection Policies are required, the broker may send back the Intune MAM Resource token
+        NSMutableDictionary *userInfo = [[NSMutableDictionary alloc] initWithDictionary:queryParamsMap];
+        if (queryParamsMap[BROKER_INTUNE_HASH_KEY] && queryParamsMap[BROKER_INTUNE_RESPONSE_KEY])
+        {
+            ADAuthenticationError *intuneTokenError = nil;
+            NSDictionary *responseDictionary = @{BROKER_RESPONSE_KEY:queryParamsMap[BROKER_INTUNE_RESPONSE_KEY],
+                                                 BROKER_HASH_KEY:queryParamsMap[BROKER_INTUNE_HASH_KEY],
+                                                 BROKER_MESSAGE_VERSION:queryParamsMap[BROKER_MESSAGE_VERSION] ? queryParamsMap[BROKER_MESSAGE_VERSION] : @1};
+
+            NSDictionary *intuneTokenResponse = [ADBrokerKeyHelper decryptBrokerResponse:responseDictionary
+                                                                   correlationId:correlationId
+                                                                           error:&intuneTokenError];
+
+            ADAuthenticationResult *intuneTokenResult = [[ADTokenCacheItem new] processTokenResponse:intuneTokenResponse
+                                                                                         fromRefresh:NO
+                                                                                requestCorrelationId:intuneTokenResponse[OAUTH2_CORRELATION_ID_RESPONSE]];
+
+            if (!keychainGroup)
+            {
+                AD_LOG_WARN(correlationId, @"Failed to cache Intune token, no resume state found in NSUserDefaults");
+            }
+            else
+            {
+                ADTokenCacheAccessor *cacheAccessor = [[ADTokenCacheAccessor alloc] initWithDataSource:[ADKeychainTokenCache keychainCacheForGroup:keychainGroup]
+                                                                                             authority:intuneTokenResult.tokenCacheItem.authority];
+
+                [cacheAccessor updateCacheToResult:intuneTokenResult cacheItem:nil refreshToken:nil context:nil];
+            }
```

Comment:
We'd better call `updateCacheToResult` only if `AD_SUCCEEDED == intuneTokenResult.status`, one condition might need to be added :)